### PR TITLE
Setup Yarn workspace

### PR DESCRIPTION
**Pull request by SamChou19815**
### Summary 

Setup Yarn workspace to enable more consistent policy enforcement and encourage better sharing of code between `frontend` and `functions`.

Blocking #313, #333, #372.

#### Rationale

Here are several advantages of Yarn workspace:

- Yarn workspace hoists the `node_modules` to the root samwise folder, so that duplicate modules won't be added to `frontend` and `functions`'s `node_modules` folder twice.
- Yarn workspace encourages code sharing. You can easily make one workspace depends on another workspace without publishing anything as npm package, and without doing any symlink magic.
- Yarn workspace is designed for monorepo, which fits our use case.
- You can continue use all your familiar CLI commands. For example, run `yarn start` in `frontend` folder will be translated to `yarn workspace frontend start` call by Yarn automatically.

Here are some additional advantages for managing repos:

- We can setup ESLint as the root of the project, and make it check all code. No more potential eslint config duplication.
- When we add prettier later, prettier can also be configured globally in the repo.

#### This PR

What this PR does:

- [x] Setup Yarn workspace
- [x] Unified our linter config
- [x] Make our CI workflows more suitable for monorepo by path filters
- [x] Suppress some linter errors to unblock this effort

In the end, we are able to partially address #367. More work needs to be done to actually reduce code duplication.

What this PR does not do and I plan to do in future PR:

- Refactor the `functions` code to share more code with `frontend`
- Fix all linter problems as a result of this change

#### Relevant Docs

- https://yarnpkg.com/en/docs/workspaces
- https://yarnpkg.com/en/docs/cli/workspace

### Test Plan 

CI jobs can still run without error.

### Notes

When this PR is merged, everyone should:

1. Delete `node_modules` and `yarn.lock` in `frontend` and `functions` folder.
2. Run `yarn` in the root folder
3. Restart IDE
4. Profit!

When this PR is merged, we need to make `lint` an required check in branch protection.

### CODE REVIEW

File: .github/workflows/cd-workflow.yml [hunk 1]
```diff
@@ -11,14 +11,19 @@ jobs:
       - uses: actions/checkout@master
       - name: Set up Node
         uses: actions/setup-node@v1
+      - name: Use Yarn Cache
+        uses: actions/cache@v1
+        with:
+          path: ~/.cache/yarn
+          key: yarn-${{ hashFiles(format('{0}{1}', github.workspace, '/yarn.lock')) }}
+          restore-keys: yarn-
```
Comment (SamChou19815) [inline .github/workflows/cd-workflow.yml]: We are now using cache in CI!

File: .github/workflows/ci-frontend.yml [hunk 2]
```diff
@@ -1,38 +1,27 @@
...
-        run: cd frontend && yarn install
+        run: yarn install
       - name: Run TypeScript Compiler
-        run: cd frontend && yarn tsc
-      - name: Run Linter
-        run: cd frontend && yarn lint
+        run: yarn workspace frontend tsc
       - name: Run Tests
-        run: cd frontend && yarn test
+        run: yarn workspace frontend test
```
Comment (SamChou19815) [inline .github/workflows/ci-frontend.yml]: See https://yarnpkg.com/en/docs/cli/workspace

File: .github/workflows/ci-lint.yml [hunk 3]
```diff
@@ -0,0 +1,22 @@
+name: CI
+on: push
```
Comment (SamChou19815) [inline .github/workflows/ci-lint.yml]: Now that we unified our linter config, we can run linter globally.

File: course-info/fetch-exam.js [hunk 4]
```diff
@@ -34,12 +34,12 @@ function parsePrelimLine(line) {
     subject,
     courseNumber,
     sectionNumber: '',
-    time: date.getTime()
+    time: date.getTime(),
```
Comment (SamChou19815) [inline course-info/fetch-exam.js]: Now `course-info` folder is also covered by the linter.

File: frontend/package.json [hunk 5]
```diff
@@ -34,30 +33,16 @@
...
-    "@typescript-eslint/parser": "^2.3.1",
-    "eslint": "^6.4.0",
-    "eslint-config-airbnb": "^18.0.1",
-    "eslint-config-react-app": "^5.0.2",
-    "eslint-plugin-import": "^2.18.2",
-    "eslint-plugin-jest": "^22.17.0",
-    "eslint-plugin-jsx-a11y": "^6.2.3",
-    "eslint-plugin-react": "^7.14.3",
-    "firebase-tools": "^7.4.0",
-    "jest": "^24.9.0",
```
Comment (SamChou19815) [inline frontend/package.json]: Moved to the root `package.json`!

File: frontend/src/components/TitleBar/Tags/ClassTagAdder.tsx [hunk 6]
```diff
@@ -18,7 +18,7 @@ type SimpleCourse = {
   readonly noSpaceName: string;
 };
 
-type Props = { readonly fuse: Fuse<SimpleCourse> | null };
+type Props = { readonly fuse: Fuse<SimpleCourse, FuseOptions<SimpleCourse>> | null };
```
Comment (SamChou19815) [inline frontend/src/components/TitleBar/Tags/ClassTagAdder.tsx]: This big change accidentally triggers some version bump. Fixed TS error as a result.

**Comment by SamChou19815** [comment]
@mt-xing Cleaned up the PR text and added relevant tasks.

**Comment by SamChou19815** [comment]
My plan to modify the CI workflow once this is adopted:

Checking on every commit job:

```yml
# ci-frontend-check.yml
name: CI Frontend
on: push to frontend and related files

jobs:
  - check: run tsc and test

# ci-functions-check.yml
name: CI Functions
on: push to functions and related files

jobs:
  - check: run tsc and test (no tests yet :()
```

PR only job:
```yml
# ci-frontend-build.yml
name: CI Frontend
on: push to frontend and related files

jobs:
  - build: build frontend
  - deploy: deploy each PR
```

With these workflows: we can make the follow jobs a required check in branch protections:

- lint
- build
- check

Except that we might need to create a dummy workflow that is like
```yml
name: Dummy
on: push

jobs:
  check:
    runs-on: ubuntu-latest
    steps:
      - name: Do nothing
        run: exit 0
  build:
    runs-on: ubuntu-latest
    steps:
      - name: Do nothing
        run: exit 0
```
The reason is that optional required checks are not supported by GitHub yet, and `check` or `build` job is not guaranteed to be triggered every time. I mentioned the hack in this post: https://github.community/t5/GitHub-Actions/Github-actions-and-required-checks-in-a-monorepo-using-paths-to/m-p/34829 GitHub still has not addressed the problem yet :(

**Comment by ptwu** [comment]
signing in on the staging site isn't working for me... is this expected?

**Comment by SamChou19815** [comment]
@ptwu signing on staging works for me. Not sure what happened.

**Pull request merged**